File: src/test/java/com/github/comius/reals/IntervalTest.java
package com.github.comius.reals;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import com.github.comius.RoundingContext;
import com.github.comius.floats.Floats;
import com.github.comius.floats.Floats.Float;
import com.github.comius.reals.Interval;

public class IntervalTest {

    Interval a = new Interval(Floats.impl().valueOf(0), Floats.impl().valueOf(1));
    Interval b = new Interval(Floats.impl().valueOf(5), Floats.impl().valueOf(6));
    RoundingContext r = new RoundingContext(0, 10);

    private void assertIn(Float a, Interval i) {
	assertTrue(a + " not in " + i, a.compareTo(i.d()) >= 0);
	assertTrue(a + " not in " + i, i.u().compareTo(a) >= 0);
    }

    private void assertOut(Float a, Interval i) {
	assertFalse(a + " in " + i, a.compareTo(i.d()) >= 0 && i.u().compareTo(a) >= 0);
    }

    @Test
    public void testAddition() {
	testAddition(a, b, r);
    }

    public void testAddition(Interval a, Interval b, RoundingContext r) {
	// Add intervals
	Interval c = a.add(b, r);

	// Verify endpoints are in the interval
	assertIn(a.d().add(b.u(), r.down), c);
	assertIn(a.u().add(b.u(), r.up), c);

	// Verify midpoints are in the interval

	Float am = a.d().split(a.u());
	Float bm = b.d().split(b.u());
	assertIn(am.add(bm, r.down), c);

	// Verify a lower endpoint - 1 is out, and upper endpoint + 1 is out assertOut(a.x().add(b.x(),
	assertOut(a.d().add(b.u(), r.down).subtract(Floats.impl().valueOf(1), r.down), c);
	assertOut(a.d().add(b.u(), r.up).add(Floats.impl().valueOf(1), r.up), c);
    }

}

File: src/test/scala/com/github/comius/reals/IntervalTest.scala
/*
 * Dedekind Reals - Java Library for computing with Dedekind Reals
 * Copyright (c) 2019 Ivo List
 *
 * This software is distributed under the terms found
 * in file LICENSE.txt that is included with this distribution.
 */

package com.github.comius.reals

import org.junit.Test
import com.github.comius.floats.Floats.{ impl => D }
import org.junit.Assert.assertEquals
import com.github.comius.RoundingContext

class IntervalTest {
  private[this] def e(d: D.T): Interval = {
    Interval(d, d)
  }

  /**
   * Tests multiplication of 0,+/- 1 with +/- inf and +/- inf with +/- inf using intervals.
   */
  @Test
  def testMultiplication(): Unit = {
    val precision = 10
    val r = new RoundingContext(0, precision)

    assertEquals(e(D.posInf), e(D.posInf).multiply(e(D.posInf), r))
    assertEquals(e(D.negInf), e(D.posInf).multiply(e(D.negInf), r))
    assertEquals(e(D.negInf), e(D.negInf).multiply(e(D.posInf), r))
    assertEquals(e(D.posInf), e(D.negInf).multiply(e(D.negInf), r))
    assertEquals(e(D.posInf), e(D.posInf).multiply(e(D.ONE), r))
    assertEquals(e(D.negInf), e(D.posInf).multiply(e(D.ONE.negate), r))
    assertEquals(e(D.negInf), e(D.negInf).multiply(e(D.ONE), r))
    assertEquals(e(D.posInf), e(D.negInf).multiply(e(D.ONE.negate), r))
    assertEquals(Interval(D.negInf, D.posInf), e(D.negInf).multiply(e(D.ZERO), r))
    assertEquals(Interval(D.negInf, D.posInf), e(D.posInf).multiply(e(D.ZERO), r))
    assertEquals(Interval(D.negInf, D.posInf), e(D.ZERO).multiply(e(D.negInf), r))
    assertEquals(Interval(D.negInf, D.posInf), e(D.ZERO).multiply(e(D.posInf), r))
  }
}
